src/hotspot/share/opto/rangeinference.hpp

```diff
@@ -0,0 +1,179 @@
+ * be 0 in v, and any bit that is 1 in ones must be 1 in v.
+ *
+ * I.e, for each bit position from 0 to sizeof(U) - 1, the corresponding bits
+ * of zeros, ones and the allowed bit in v must follow:
+ *
+ * zeros    ones    allowed bits
+ * 0        0       0 or 1
+ * 1        0       0
+ * 0        1       1
+ * 1        1       none (impossible state)
+ *
+ * E.g:
+ * zeros: 00110100
+ * ones:  10000010
+ * Then:  10001010 would satisfy the bit constraints
+ * while: 10011000 would not since the bit at the 4th position violates
+ * zeros and the bit at the 7th position violates ones
+ *
+ * A KnownBits is sane if there is no position at which a bit must be both set
+ * and unset at the same time. That is (zeros & ones) == 0.
+ */
+template <class U>
+class KnownBits {
+  static_assert(std::is_unsigned<U>::value, "bit info should be unsigned");
+
+public:
+  U _zeros;
+  U _ones;
+
+  bool is_satisfied_by(U v) const {
+    return (v & _zeros) == 0 && (v & _ones) == _ones;
+  }
+};
+
+// All the information needed to construct a TypeInt/TypeLong, the constraints
+// here may be arbitrary and need to be canonicalized to construct a
+// TypeInt/TypeLong
+template <class S, class U>
+class TypeIntPrototype {
+public:
+  static_assert(std::is_signed<S>::value, "");
+  static_assert(std::is_unsigned<U>::value, "");
+  static_assert(sizeof(S) == sizeof(U), "");
+
+  RangeInt<S> _srange;
+  RangeInt<U> _urange;
+  KnownBits<U> _bits;
+
+private:
+  friend class TypeInt;
+  friend class TypeLong;
+
+  template <class T1, class T2>
+  friend void test_canonicalize_constraints_simple();
+
+  template <class T1, class T2>
+  friend void test_canonicalize_constraints_random();
+
+  // A canonicalized version of a TypeIntPrototype, if the prototype represents
+  // an empty type, _present is false, otherwise, _data is canonical
+  class CanonicalizedTypeIntPrototype {
+  public:
+    bool _present; // whether this is an empty set
+    TypeIntPrototype<S, U> _data;
+
+    static CanonicalizedTypeIntPrototype make_empty() {
+      return {false, {}};
+    }
+  };
+
+  CanonicalizedTypeIntPrototype canonicalize_constraints() const;
+  int normalize_widen(int w) const;
+#ifdef ASSERT
+  bool contains(S v) const;
+  void verify_constraints() const;
+#endif // ASSERT
+};
+
+// Various helper functions for TypeInt/TypeLong operations
+class TypeIntHelper {
+public:
+  // Calculate the cardinality of a TypeInt/TypeLong ignoring the bits
+  // constraints, the result is tuned down by 1 to ensure the bottom type is
+  // correctly calculated
+  template <class S, class U>
+  static U cardinality_from_bounds(const RangeInt<S>& srange, const RangeInt<U>& urange) {
+    static_assert(std::is_signed<S>::value, "");
+    static_assert(std::is_unsigned<U>::value, "");
+    static_assert(sizeof(S) == sizeof(U), "");
+
+    if (U(srange._lo) == urange._lo) {
+      return urange._hi - urange._lo;
+    }
```

Comment:
Here you check that the two ranges are identical? Maybe add an assert that also the `hi` limit is identical? Or did I get something wrong?